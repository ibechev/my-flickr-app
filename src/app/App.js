import React, { useEffect } from "react";
import PropTypes from "prop-types";
import { connect } from "redhooks";
import { fetchImagesAction } from "../app/actions/index";
import {
  START_FETCHING_MORE,
  FETCHING_MORE_COMPLETE,
  FETCH_IMAGES_ERROR
} from "./actions/types";

import SearchBoxContainer from "./components/searchBox/SearchBoxContainer";
import ImageGallery from "./components/imageGallery/ImageGallery";
import LoadingSearch from "./components/loadingSearch/LoadingSearch";
import IconDialog from "./components/iconDialog/IconDialog";

const App = ({
  isFetchingMore,
  fetchImagesAction,
  noMorePages,
  images,
  searching,
  noResults,
  location,
  history
}) => {
  useEffect(() => {
    window.addEventListener("scroll", detectScrollEnd, false);

    return function clearEventListener() {
      window.removeEventListener("scroll", detectScrollEnd, false);
    };
  });

  const detectScrollEnd = () => {
    const galleryH = document.getElementById("image-gallery").clientHeight;
    const searchBoxH = document.getElementById("search-box").clientHeight;
    const offsetFetchTrigger = galleryH + searchBoxH - window.innerHeight;

    if (
      window.pageYOffset > offsetFetchTrigger &&
      !isFetchingMore &&
      !noMorePages &&
      !!images.length
    ) {
      window.removeEventListener("scroll", detectScrollEnd, false);
      fetchImagesAction({
        type: START_FETCHING_MORE,
        meta: {
          success: FETCHING_MORE_COMPLETE,
          error: FETCH_IMAGES_ERROR
        }
      });
    }
  };

  return (
    <div className="app">
      <SearchBoxContainer history={history} location={location} />

      <ImageGallery />

      {searching && <LoadingSearch />}

      {!images.length && !searching && !noResults && (
        <IconDialog icon="images-placeholder" text="Search images on Flickr" />
      )}

      {noResults && !searching && (
        <IconDialog icon="no-results" text="Oops, nothing found." />
      )}
    </div>
  );
};

const mapStateToProps = ({
  isFetchingMore,
  noMorePages,
  images,
  searching,
  noResults
}) => ({
  isFetchingMore,
  noMorePages,
  images,
  searching,
  noResults
});

App.propTypes = {
  isFetchingMore: PropTypes.bool.isRequired,
  fetchImagesAction: PropTypes.func.isRequired,
  noMorePages: PropTypes.bool.isRequired,
  images: PropTypes.array,
  searching: PropTypes.bool.isRequired,
  noResults: PropTypes.bool.isRequired,
  location: PropTypes.object.isRequired,
  history: PropTypes.object.isRequired
};

export default connect(
  mapStateToProps,
  { fetchImagesAction }
)(App);
