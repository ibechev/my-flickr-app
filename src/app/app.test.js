import React from "react";
import renderer from "react-test-renderer";
import App from "./App";
import { mount } from "enzyme";
import { image, testTags } from "../../tests/testValues";

describe("Expect <App />", () => {
  const defaultState = {
    page: 1,
    loading: false,
    fetchingMore: false,
    NoMorePages: false,
    noResults: false,
    images: [],
    tags: []
  };

  const setup = (stateTestProperties = {}) => {
    const wrapper = mount(<App />);
    wrapper.setState({ ...defaultState, ...stateTestProperties });
    return wrapper;
  };

  it.only("to render correctly", () => {
    // const wrapper = renderer.create(<App />).toJSON();
    // expect(wrapper).toMatchSnapshot();
  });

  it("to only display <ImageGallery />, if 'state.loading' is 'false' and 'state.noResults' is 'false'", () => {
    const wrapperOne = setup({ loading: true });
    expect(wrapperOne.find("ImageGallery")).toHaveLength(0);

    const wrapperTwo = setup({ noResults: true });
    expect(wrapperTwo.find("ImageGallery")).toHaveLength(0);

    const wrapperThree = setup({ loading: true, noResults: true });
    expect(wrapperThree.find("ImageGallery")).toHaveLength(0);

    const wrapper = setup(defaultState);
    expect(wrapper.find("ImageGallery")).toHaveLength(1);
  });

  it("to only display <span class='spinner-main' />, if 'state.loading' equals 'true'", () => {
    const wrapperOne = setup({ loading: true });
    expect(wrapperOne.find("span.spinner-main")).toHaveLength(1);

    const wrapper = setup();
    expect(wrapper.find("span.spinner-main")).toHaveLength(0);
  });

  it("to only display <span class='spinner-scroll' />, if 'state.images' is not empty, 'state.loading' is 'false', 'state.noResults' is 'false' and 'state.noMorePages' is 'false'", () => {
    const wrapperOne = setup({
      images: [image, image, image]
    });
    expect(wrapperOne.find("span.spinner-scroll")).toHaveLength(1);

    const wrapperTwo = setup({
      images: [image, image, image],
      loading: true
    });
    expect(wrapperTwo.find("span.spinner-scroll")).toHaveLength(0);

    const wrapperThree = setup({
      images: [image, image, image],
      noResults: true
    });
    expect(wrapperThree.find("span.spinner-scroll")).toHaveLength(0);

    const wrapperFour = setup({
      images: [image, image, image],
      noMorePages: true
    });
    expect(wrapperFour.find("span.spinner-scroll")).toHaveLength(0);
    const wrapperFive = setup({
      images: [image, image, image],
      noMorePages: true
    });
    expect(wrapperFive.find("span.spinner-scroll")).toHaveLength(0);

    const wrapperSix = setup({
      loading: true,
      noResults: true
    });
    expect(wrapperSix.find("span.spinner-scroll")).toHaveLength(0);

    const wrapperSeven = setup({
      loading: true,
      noResults: true,
      noMorePages: true
    });
    expect(wrapperSeven.find("span.spinner-scroll")).toHaveLength(0);

    const wrapperEight = setup({
      noResults: true,
      noMorePages: true
    });
    expect(wrapperEight.find("span.spinner-scroll")).toHaveLength(0);

    const wrapperNine = setup({
      loading: true,
      noMorePages: true
    });
    expect(wrapperNine.find("span.spinner-scroll")).toHaveLength(0);
  });

  it("to only display <div class='images-placeholder' />, if 'state.images' is empty, 'state.loading' is 'false' and 'state.noResults' is 'false'", () => {
    const wrapper = setup();
    expect(wrapper.find("div.images-placeholder")).toHaveLength(1);

    const wrapperOne = setup({ images: [image, image, image] });
    expect(wrapperOne.find("div.images-placeholder")).toHaveLength(0);

    const wrapperTwo = setup({
      images: [image, image, image],
      loading: true
    });
    expect(wrapperTwo.find("div.images-placeholder")).toHaveLength(0);

    const wrapperThree = setup({
      images: [image, image, image],
      loading: true,
      noResults: true
    });
    expect(wrapperThree.find("div.images-placeholder")).toHaveLength(0);

    const wrapperFour = setup({
      images: [image, image, image],
      noResults: true
    });
    expect(wrapperFour.find("div.images-placeholder")).toHaveLength(0);

    const wrapperFive = setup({
      noResults: true
    });
    expect(wrapperFive.find("div.images-placeholder")).toHaveLength(0);

    const wrapperSix = setup({
      loading: true
    });
    expect(wrapperSix.find("div.images-placeholder")).toHaveLength(0);

    const wrapperSeven = setup({
      loading: true,
      noResults: true
    });
    expect(wrapperSeven.find("div.images-placeholder")).toHaveLength(0);
  });
  it("to only display <div class='no-results-placeholder' />, if 'state.noResults' is 'true' and 'state.loading' is 'false'", () => {
    const wrapper = setup({ noResults: true });
    expect(wrapper.find("div.no-results-placeholder")).toHaveLength(1);

    const wrapperOne = setup({ noResults: true, loading: true });
    expect(wrapperOne.find("div.no-results-placeholder")).toHaveLength(0);

    const wrapperThree = setup({ loading: true });
    expect(wrapperThree.find("div.no-results-placeholder")).toHaveLength(0);
  });
});
