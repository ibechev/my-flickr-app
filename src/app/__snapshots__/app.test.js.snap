// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Expect <App /> to render correctly 1`] = `
<div
  className="app"
>
  <h1>
    Hey hey hey
  </h1>
  <hr />
  <ul>
    <li>
      Express
    </li>
    <li>
      React
    </li>
    <li>
      Webpack
    </li>
    <li>
      HMR
    </li>
    <li>
      SCSS
    </li>
  </ul>
</div>
`;
